### PR TITLE
Extract competitor limit validator

### PR DESCRIPTION
**Pull request by viroulep**

### CODE REVIEW

File: WcaOnRails/spec/lib/results_validators/competitor_limit_validator_spec.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      }
+    }
+
+    context "for competitions having a competitor limit" do
+      before(:example) do
+        [Result, InboxResult].each do |model|
+          result_kind = model.model_name.singular.to_sym
+          10.times {
+            FactoryBot.create(result_kind, competition: competition1, eventId: "333oh")
+            FactoryBot.create(result_kind, competition: competition2, eventId: "222")
```
Comment (jonatanklosko) [inline WcaOnRails/spec/lib/results_validators/competitor_limit_validator_spec.rb]: Perhaps we could use `FactoryBot.create_list(result_kind, 10, ...)` here?

Comment (viroulep) [reply WcaOnRails/spec/lib/results_validators/competitor_limit_validator_spec.rb]: Yes absolutely!

File: WcaOnRails/spec/factories/inbox_persons.rb [hunk 2]
```diff
@@ -6,9 +6,9 @@
     # and id is not the usual auto increment integer (it's actually a varchar!)
     # Therefore we make the simple choice of always setting the id based on
     # what is present in the db.
-    # 'maximum(:id).to_i' always work: either it's nil and returns 0, or it just
-    # returns the appropriate number.
-    id { (InboxPerson.maximum(:id).to_i + 1) }
+    # Since id is a varchar, `maximum(:id)` only works up to "9"...
+    # Therefore we must do the max logic in RoR's world after casting.
```
Comment (jfly) [inline WcaOnRails/spec/factories/inbox_persons.rb]: :sob:

File: WcaOnRails/lib/results_validators/competitor_limit_validator.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      " Unless a specific agreement was made when announcing the competition (such as a per-day competitor limit), the results of the competitors registered after the competitor limit was reached must be removed."
+
+    @@desc = "For competition with a competitor limit, this validator checks that this limit is respected."
+
+    def validate(competition_ids: [], model: Result, results: nil)
+      reset_state
+
+      if competition_ids.empty?
+        competition_ids = results.map(&:competitionId).uniq
+      end
```
Comment (jonatanklosko) [inline WcaOnRails/lib/results_validators/competitor_limit_validator.rb]: Ah actually, why do we set competition ids from results here, I believe in most places it's the other way round: https://github.com/thewca/worldcubeassociation.org/blob/206794e0935e18b9d7373fca8b7279bb837e7d5f/WcaOnRails/lib/results_validators/scrambles_validator.rb#L14-L15

Comment (viroulep) [reply WcaOnRails/lib/results_validators/competitor_limit_validator.rb]: We don't actually need any result for this check, we just want a list of competitions!
Either `competition_ids` are provided and we check these competitions, or some `results` are provided from which we want to extract the corresponding competitions.

Comment (jonatanklosko) [reply WcaOnRails/lib/results_validators/competitor_limit_validator.rb]: Ah, because we query results anyway, gotcha!

**Pull request merged**